sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -1150,6 +1150,23 @@ class DataFrame private[sql](
   }
 
   /**
+   * Returns a new [[DataFrame]] by adding a column with metadata.
+   */
+  private[spark] def withColumn(colName: String, col: Column, metadata: Metadata): DataFrame = {
```

Comment:
should we make `Column.as(alias: String, metadata: Metadata)` also private? As it also expose the `Metadata` too.